src/training/scheduler.h

```diff
@@ -154,11 +152,10 @@ class Scheduler : public TrainingObserver {
       : options_(options), state_(state) {
     ABORT_IF(state_->factor != 1, "state.factor unexpectedly not 1 at this point??");
     updateLearningRate(*state);
-    installSignalHandlers();
   }
 
   bool keepGoing() {
-    if(getSigtermFlag()) // received signal SIGERM => exit gracefully
+    if(getSignalFlag(SIGTERM)) // received signal SIGERM => exit gracefully
```

Comment:
I realize that there is too much duplication of knowledge on what it means to signal this job to terminate. E.g. you duplicate the knowledge that it is SIGTERM at four places up to here.

I suggest to abstract this. Have a function `terminationRequested()` or so, which encapsulates the mechanism. One can also imagine much less risky mechanisms, e.g. creating a file whose presence indicates a termination request. It is also not clear to a reader whether reading out the signal clears it or not.

Signals are risky in that they interfere with UNIX' idiotic EINTR protocol, which is, for example, not implemented correctly in some versions of NCCL, and likely not implemented correctly in any of our external file-reading/writing libraries (e.g. the NPZ writer quite surely does not follow it, and I wouldn't put it beyond Protobufs either).

Reply:
There is nothing to abstract here. We want to react to one very specific scenario: The process receives a request to save its current state and exit a.s.a.p. That's called a graceful shutdown and exactly what SIGTERM and signal handling are meant for. There are other termination requests where we do NOT want to save the state of training, e.g. after SIGINT. 

In fact, on linux, (glibc or later), if I read [this](https://man7.org/linux/man-pages/man2/signal.2.html) correctly, installing a signal handler through the gcc signal function (which wraps around sigaction) will by default restarted interruptible system calls that are interrupted by a signal. Except data transfer functions that do not return EINTR but instead the number of bytes transferred, so chances are the NPZ writer won't stumble over interrupted system calls.